### ISSUE TITLE
SpongeForge coremod loading too late to mixin to Event bus

### ISSUE THREAD

**Issue by RafaelPL** [SpongePowered/SpongeForge#511]
[17:11:51] [Server thread/ERROR]: Encountered an unexpected exception
java.lang.ClassCastException: net.minecraftforge.fml.common.eventhandler.EventBus cannot be cast to org.spongepowered.mod.interfaces.IMixinEventBus
        at org.spongepowered.mod.event.SpongeModEventManager.post(SpongeModEventManager.java:273) ~[SpongeModEventManager.class:1.8-1577-3.1.0-BETA-1045]
        at org.spongepowered.mod.event.SpongeModEventManager.post(SpongeModEventManager.java:245) ~[SpongeModEventManager.class:1.8-1577-3.1.0-BETA-1045]
        at org.spongepowered.common.SpongeImpl.postEvent(SpongeImpl.java:142) ~[SpongeImpl.class:1.8-1577-3.1.0-BETA-1045]
        at net.minecraft.server.MinecraftServer.func_71247_a(MinecraftServer.java:445) ~[MinecraftServer.class:?]
        at net.minecraft.server.dedicated.DedicatedServer.func_71197_b(DedicatedServer.java:259) ~[po.class:?]
        at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:438) [MinecraftServer.class:?]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_45]

**Comment by simon816** [comment]
Can you post a pastebin of the full `fml-server-latest.log` pleae?

**Comment by RafaelPL** [comment]
1: http://hastebin.com/doqelasoba.scala
2: http://hastebin.com/lejeqoxofe.vhdl

**Comment by Mumfrey** [comment]
This occurred with another mod, it means that another coremod is causing the forge event bus to be loaded earlier than usual, we will need to look into a way of fixing this in Mixin or by making our coremod a lower priority.

**Comment by RafaelPL** [comment]
So what now? : (

**Comment by Mumfrey** [comment]
If you want something to try in the mean time, you can try to remove coremods which are loading before sponge. eg. `1.8-MB_Battlegear2-Bullseye-1.0.10.0.jar` and `redstonepaste-mc1.8-1.7.1.jar`, but other than that you will have to wait for us to investigate the issue.

**Comment by Aaron1011** [comment]
@RafaelPL: Can you try reproducing this with each of those coremods individually, to determine which one is causing the issue?

**Comment by RafaelPL** [comment]
@Aaron1011 - 

1.8-MB_Battlegear2-Bullseye-1.0.10.0.jar: http://hastebin.com/giyezudofi.scala

**Comment by simon816** [comment]
I think I found where the problem with BattleGear2 is. They load `FMLCommonHandler` in a transformer https://github.com/Mine-and-blade-admin/Battlegear2/blob/master/battlegear%20mod%20src/minecraft/mods/battlegear2/coremod/transformers/NetServerHandlerTransformer.java#L102

**Comment by RafaelPL** [comment]
Cmon guys! Please, I want to use Sponge :(

**Comment by Mumfrey** [comment]
This is a massive issue to fix, it's not going to be resolved soon as the entire nature of launching the game needs to be revisited. Discussions are ongoing about how to proceed.

**Comment by RafaelPL** [comment]
Really guys? Really? 

(RafaelPL commented on 22 Jan)

**Comment by Mumfrey** [comment]
I will reiterate: this is very, very far from being a simple issue to fix.

**Comment by modwizcode** [comment]
Any chance a temporary workaround could be developed to patch load order for specific coremods?

**Comment by Mumfrey** [comment]
There is a temporary workaround, it was mentioned above.

**Comment by RafaelPL** [comment]
Don't say that is big problem because you haven't done anything with this for the past 2 months.

**Comment by Mumfrey** [comment]
Having to change a filename is inconvenient, a proper fix for this involves a quite significant change to some internals of FML.

**Comment by ryantheleach** [comment]
If you mean removing the mods, that was mentioned, but you stated 

> "Having to change a filename is inconvenient, " Are you insinuating there is a way to reorder the load order by changing the file names of the mods that arn't loading? If so it hasn't been communicated in this thread.

**Comment by Deamon5550** [comment]
@ryantheleach its in the referenced issue
>The simplest way of doing that would be either renaming the Sponge jar with an underscore at the beginning like _sponge.whatever.jar or renaming MovingWorld jar to zzzMovingWorld.jar.

>Basically the problem boils down to coremods being initialised in alphabetical order and not priority order, we are pursuing solutions to this internally with the relevant people.

**Comment by simon816** [comment]
As an alternative solution I have created SpongeBootstrap - a launchable jar that will inject SpongeForge into the FML loading process before any other coremods.
https://forums.spongepowered.org/t/spongebootstrap-fixes-ordering-issue-with-other-coremods/12275

**Comment by jpiermarini** [comment]
Is this why adding enderio and endercore has caused issues for me? Pex and essentialcmds don't work when those two are loaded.

**Comment by MarkL4YG** [comment]
So, why not just temporarily rename the download to begin with _ until there's a fix ?

**Comment by Aetheros** [comment]
@MarkL4YG For the same reason why doctors don't give patients pain killers untill a cure for cancer is found. Fixing the problem with a simple, but incredibly incomplete workaround does only mask the symptoms, but doesn't fix the problem, which remove the immediate need for a fix, slowing down development on the fix for said problem, while it also could mask the symptoms of other, equally or more serious threads to the health of sponge, which would only be found once the fix for the current issue is developed.

**Comment by ryantheleach** [comment]
Whilst your comment was correct Aetheros that was a very confusing analogy.

**Comment by kinggoesgaming** [comment]
It's a good analogy but not worded right :P

**Comment by Aetheros** [comment]
I'm not a native speaker, so please forgive my wording D:

**Comment by MarkL4YG** [comment]
xD

**Comment by Zidane** [comment]
Closing this as it will eventually be resolved once cpw from Forge Team implements the new platform loading functionality to LaunchWrapper. At that point, we can finally be able to do what we need to do before coremods are even loaded.

Let me stress that this isn't a Sponge issue, it is an issue with how Sponge is treated by FML during the loading process. SpongeForge is mod in name only, meaning that while we are displayed as a mod to the masses, we are like no mod in existence; we change as much code as Forge does at this point. In the spirit of working in-hand with Forge mods, we subject ourselves to the same loading process that other coremods do which causes the issues seen above. Once the new system is finished, we will have support for letting us do what we need to do before coremods interfere.

On a final note, the coremod in question is classloading a class in their coremod that they should NOT be doing so they are at fault to begin with..
